pkg/epp/datastore/datastore.go

```diff
@@ -126,6 +126,9 @@ func (ds *datastore) PoolHasSynced() bool {
 func (ds *datastore) PoolLabelsMatch(podLabels map[string]string) bool {
 	ds.poolAndModelsMu.RLock()
 	defer ds.poolAndModelsMu.RUnlock()
+	if ds.pool == nil {
```

Comment:
that's good to have without any relation to the predicate.
nit: you can use the internal ds function (same logic, just to be consistent with GetPool func)
```
if !ds.PoolHasSynced() {
	return false
}
```

Reply:
It will cause deadlock if we use `ds.PoolHasSynced` because it also need to acquire `poolAndModelsMu`.

Reply:
are you sure? :)
Go mutex behavior around reentrant locks can be a bit nuanced. 
see this example here https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/epp/datastore/datastore.go#L112-L117.
you can give it a try in your fork.

Reply:
I think it's safe not to acquire RWMutex twice in one function.
https://pkg.go.dev/sync#RWMutex
> If any goroutine calls [RWMutex.Lock](https://pkg.go.dev/sync#RWMutex.Lock) while the lock is already held by one or more readers, concurrent calls to [RWMutex.RLock](https://pkg.go.dev/sync#RWMutex.RLock) will block until the writer has acquired (and released) the lock, to ensure that the lock eventually becomes available to the writer. Note that this prohibits recursive read-locking.

https://stackoverflow.com/questions/30547916/goroutine-blocks-when-calling-rwmutex-rlock-twice-after-an-rwmutex-unlock

Reply:
it's definitely safe to keep current check.
was just a nice to have comment :)